node/src/tasks/msp_upload_file.rs

```diff
@@ -0,0 +1,854 @@
+            .await
+            .map_err(|e| {
+                let err_msg = format!("Failed to query available storage capacity: {:?}", e);
+                error!(
+                    target: LOG_TARGET,
+                    err_msg
+                );
+                anyhow::anyhow!(err_msg)
+            })?;
+
+        // Increase storage capacity if the available capacity is less than the file size.
+        if available_capacity < event.size {
+            warn!(
+                target: LOG_TARGET,
+                "Insufficient storage capacity to accept file: {:?}",
+                event.file_key
+            );
+
+            let current_capacity = self
+                .storage_hub_handler
+                .blockchain
+                .query_storage_provider_capacity(own_msp_id)
+                .await
+                .map_err(|e| {
+                    let err_msg = format!("Failed to query storage provider capacity: {:?}", e);
+                    error!(
+                        target: LOG_TARGET,
+                        err_msg
+                    );
+                    anyhow::anyhow!(err_msg)
+                })?;
+
+            let max_storage_capacity = self
+                .storage_hub_handler
+                .provider_config
+                .max_storage_capacity;
+
+            if max_storage_capacity == current_capacity {
+                let err_msg = "Reached maximum storage capacity limit. Unable to add more more storage capacity.";
+                warn!(
+                    target: LOG_TARGET, err_msg
+                );
+                return Err(anyhow::anyhow!(err_msg));
+            }
+
+            let new_capacity = self.calculate_capacity(&event, current_capacity)?;
+
+            let call = storage_hub_runtime::RuntimeCall::Providers(
+                pallet_storage_providers::Call::change_capacity { new_capacity },
+            );
+
+            let earliest_change_capacity_block = self
+                .storage_hub_handler
+                .blockchain
+                .query_earliest_change_capacity_block(own_msp_id)
+                .await
+                .map_err(|e| {
+                    error!(
+                        target: LOG_TARGET,
+                        "Failed to query storage provider capacity: {:?}", e
+                    );
+                    anyhow::anyhow!("Failed to query storage provider capacity: {:?}", e)
+                })?;
+
+            // Wait for the earliest block where the capacity can be changed.
+            self.storage_hub_handler
+                .blockchain
+                .wait_for_block(earliest_change_capacity_block)
+                .await?;
+
+            self.storage_hub_handler
+                .blockchain
+                .send_extrinsic(call, Tip::from(0))
+                .await?
+                .with_timeout(Duration::from_secs(60))
+                .watch_for_success(&self.storage_hub_handler.blockchain)
+                .await?;
```

Comment:
After taking a look at the code for `wait_for_block` I'm still not totally sure so might as well ask. This task gets frozen here until the condition is met, right?
Would it be more efficient in `wait_for_block` to (instead of just blocking the task until the required block number arrives) receive a callback and calling that callback when the required block number is received? That way we could free resources quicker

Reply:
Also, if there're two of this task running, when the first one changes the capacity, the `earliest_change_capacity_block` changes but the second task never realizes that, so it'll try to send the extrinsic and it will fail. That could also be solved using the callback model. Again, this is only valid if `wait_for_block` works as I think it does 😬

Reply:
IF it does and IF we decide to change this, I don't think I'd do it in this PR, just as a FYI. @ffarall wdyt?

Reply:
`wait_for_block` does exactly that. But yes the task will wait until it is reached and it is possible to have mutliple tasks waiting waiting to add capacity. In that case we should probably check after waiting for so long if the capacity has already changed and if it covers what we needed.

Reply:
Conclusion: We actually need a unified piece of code in the blockchain service that "accumulates" all the needed storage capacity changes in the various tasks, and when we reach the block where the capacity change is allowed, it makes the capacity change for all of them considered. This would be left for another PR.